src/core/utils/CorePlugin.ts

```diff
@@ -20,7 +30,96 @@ export class CorePlugin extends JWPlugin {
+     * Render the `vDocument`.
+     */
+    render(): void {
+        this.vDocument.renderer.render(this.vDocument.editable);
+    }
+
+    //--------------------------------------------------------------------------
+    // Private
+    //--------------------------------------------------------------------------
+
+    /**
+     * Return a position in the `VDocument` as a tuple `[VNode, 'before' | 'after']`
+     * (ie. the position is before or after a given `VNode`). The position is
+     * always given on the leaf.
+     *
+     * @param container
+     * @param offset
+     */
+    _getPosition(container: DOMElement | Node, offset: number): VPosition {
+        const containers = this.vDocument.domMap.get(container);
+        const isTextNode = container.nodeType === 3;
+        if (isTextNode) {
+            return this._positionToLeaf({
+                ref: containers[offset] || containers[containers.length - 1],
+                side: offset < containers.length ? 'before' : 'after',
+            });
+        }
+        return this._positionToLeaf({ ref: containers[offset], side: 'before' });
+    }
+    /**
+     * Return a tuple with the range nodes in ltr order (rtl if `reverse` is
+     * true) and updates their `order` property accordingly.
+     *
+     * @param reverse
+     */
+    _orderedRangeNodes(reverse = false): [VNode, VNode] {
+        const rangeNodes = this.vDocument.rangeNodes;
+        const res: [VNode, VNode] = [rangeNodes.start, rangeNodes.end];
+        if (reverse) {
+            res.reverse();
+        }
+        res.forEach((rangeNode: VNode, i: 0 | 1) => {
+            rangeNode.order = i;
+        });
+        return res;
+    }
+    /**
+     * Move the position to its deepest first descendent.
+     *
+     * @param position
+     */
+    _positionToLeaf(position: VPosition): VPosition {
+        while (position.ref.children.length) {
+            position.ref = position.ref.firstChild;
+            position.side = 'before';
+        }
+        return position;
+    }
+    /**
+     * Move the `Range VNodes` to the position in `VDocument` corresponding to
+     * the given DOM Range and trigger a rendering.
+     *
+     * @param range
+     */
+    _setRange(range: Range): void {
+        const startPos: VPosition = this._getPosition(range.startContainer, range.startOffset);
+        const endPos: VPosition = this._getPosition(range.endContainer, range.endOffset);
+        const [first, last] = this._orderedRangeNodes(range['direction'] === 'rtl');
+
+        // Move the range nodes. If the start has to be moved *after* its
+        // reference node, the end should be moved first.
+        if (startPos.side === 'after') {
+            endPos.ref[endPos.side](last);
+            startPos.ref[startPos.side](first);
+        } else {
+            startPos.ref[startPos.side](first);
+            endPos.ref[endPos.side](last);
+        }
+        this.dispatcher.dispatch(this.intent({ name: 'render' }));
```

Comment:
I don't think plugins should be able to trigger a re-render (at least not until we face a true need for it.) IMO, the dispatcher calls the handlers for an intent, when the handlers have been called (potentially recursively) and they all got their chance at modifying the virtual document, the dispatcher can then trigger the commitState and the rendering process.

Reply:
This should indeed not remain once we have a mechanism for "committing" the changes to the `VDocument`, which depends on CHM's memory system. But indeed I can move this to the dispatcher, at the end of the loop, though right now we have no mechanism to ensure that the flow is over and we don't need to run the loop again.

```typescript
// In `Dispatcher`:
if (Object.keys(handlers).length && action.name !== 'render') {
    this.dispatch(ActionGenerator.intent({ name: 'render' }));
} else {
        console.warn(`No plugin is listening to the ${action.type} "${action.name}".`);
}
```